chia/wallet/wallet.py

```diff
@@ -276,16 +276,70 @@ async def select_coins(self, amount, exclude: List[Coin] = None) -> Set[Coin]:
             self.id()
         )
+        # check for exact 1 to 1 coin match.
         for coinrecord in unspent:
-            if sum_value >= amount and len(used_coins) > 0:
-                break
             if coinrecord.coin.name() in unconfirmed_removals:
                 continue
             if coinrecord.coin in exclude:
                 continue
-            sum_value += coinrecord.coin.amount
-            used_coins.add(coinrecord.coin)
-            self.log.debug(f"Selected coin: {coinrecord.coin.name()} at height {coinrecord.confirmed_block_height}!")
+            if coinrecord.coin.amount == amount:
+                sum_value += coinrecord.coin.amount
+                used_coins.add(coinrecord.coin)
+                self.log.debug(
+                    f"Selected coin: {coinrecord.coin.name()} at height {coinrecord.confirmed_block_height}!"
+                )
+                break
+
+        # Check for exact match with all coins smaller than the amount.
+        # If the total of all the smaller coins are less than the amount we choose; the smallest coin available.
+        # If we have more, smaller coins than the amount we run the next algorithm.
+        if len(used_coins) == 0:
+            smaller_coins: Set = set()
+            for coinrecord in unspent:
+                if coinrecord.coin.name() in unconfirmed_removals:
+                    continue
+                if coinrecord.coin in exclude:
+                    continue
+                if coinrecord.coin.amount < amount:
+                    sum_value += coinrecord.coin.amount
+                    smaller_coins.add(coinrecord.coin)
+            if sum_value < amount:
+                smallest_value = float("inf")  # smallest coins value
+                # for edge case where the coins are excluded or not confirmed yet.
+                smallest_coin: Optional[WalletCoinRecord] = None
+                for coinrecord in unspent:
+                    if coinrecord.coin.name() in unconfirmed_removals:
+                        continue
+                    if coinrecord.coin in exclude:
+                        continue
+                    if amount < coinrecord.coin.amount < smallest_value:
+                        # try to find a coin that is as close as possible to the amount.
+                        smallest_value = coinrecord.coin.amount
+                        smallest_coin = coinrecord
+                if smallest_coin:
+                    used_coins.add(smallest_coin.coin)
+                    sum_value = smallest_coin.coin.amount
+                    self.log.debug(
+                        f"Selected coin: {smallest_coin.coin.name()} at height {smallest_coin.confirmed_block_height}!"
+                    )
+            elif sum_value == amount:
+                used_coins = smaller_coins
+
+        if len(used_coins) == 0:
+            sum_value = 0
+            # check for all other combinations of coins.
```

Comment:
It's not checking all combinations, it's adding one at a time until they are sufficient right? In what order are the coins sorted in? It's worth a comment because that is very important here. 

Is this the same as bitcoins algorithm?

Also probably should decompose this out into it's own little module.

Reply:
@mariano54 that uses the same code as before.